src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const transformedState = structuredClone(state);
+  const stateHistory = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    if (actions[i].type === 'addProperties') {
+      for (const key in actions[i].extraData) {
+        transformedState[key] = actions[i].extraData[key];
+      }
+    } else if (actions[i].type === 'removeProperties') {
+      for (const key of actions[i].keysToRemove) {
+        delete transformedState[key];
+      }
+    } else {
+      for (const key in transformedState) {
+        delete transformedState[key];
+      }
```

Comment:
The 'clear' action type is not explicitly checked. It's assumed to be the 'else' case, which could lead to unexpected behavior if a new action type is introduced without updating this code. It's better to explicitly check for the 'clear' action type. Additionally, there should be a default case in the switch statement for error handling, as per checklist items 2 and 3.